chainerrl/agents/td3.py

```diff
@@ -0,0 +1,374 @@
+
+    def batch_select_onpolicy_action(self, batch_obs):
+        with chainer.using_config('train', False), chainer.no_backprop_mode():
+            batch_xs = self.batch_states(batch_obs, self.xp, self.phi)
+            batch_action = self.policy(batch_xs).sample().array
+        return list(cuda.to_cpu(batch_action))
+
+    def batch_act(self, batch_obs):
+        return self.batch_select_onpolicy_action(batch_obs)
+
+    def batch_act_and_train(self, batch_obs):
+        """Select a batch of actions for training.
+
+        Args:
+            batch_obs (Sequence of ~object): Observations.
+
+        Returns:
+            Sequence of ~object: Actions.
+        """
+
+        if (self.burnin_action_func is not None
+                and self.policy_optimizer.t == 0):
+            batch_action = [self.burnin_action_func()
+                            for _ in range(len(batch_obs))]
+        else:
+            batch_onpolicy_action = self.batch_select_onpolicy_action(
+                batch_obs)
+            batch_action = [
+                self.explorer.select_action(
+                    self.t, lambda: batch_onpolicy_action[i])
+                for i in range(len(batch_onpolicy_action))]
+
+        self.batch_last_obs = list(batch_obs)
+        self.batch_last_action = list(batch_action)
+
+        return batch_action
+
+    def batch_observe_and_train(
+            self, batch_obs, batch_reward, batch_done, batch_reset):
+        for i in range(len(batch_obs)):
+            self.t += 1
+            if self.batch_last_obs[i] is not None:
+                assert self.batch_last_action[i] is not None
+                # Add a transition to the replay buffer
+                self.replay_buffer.append(
+                    state=self.batch_last_obs[i],
+                    action=self.batch_last_action[i],
+                    reward=batch_reward[i],
+                    next_state=batch_obs[i],
+                    next_action=None,
+                    is_state_terminal=batch_done[i],
+                    env_id=i,
+                )
+                if batch_reset[i] or batch_done[i]:
+                    self.batch_last_obs[i] = None
+                    self.replay_buffer.stop_current_episode(env_id=i)
+            self.replay_updater.update_if_necessary(self.t)
+
+    def batch_observe(self, batch_obs, batch_reward,
+                      batch_done, batch_reset):
+        pass
+
+    def stop_episode_and_train(self, state, reward, done=False):
+
+        assert self.last_state is not None
+        assert self.last_action is not None
+
+        # Add a transition to the replay buffer
+        self.replay_buffer.append(
+            state=self.last_state,
+            action=self.last_action,
+            reward=reward,
+            next_state=state,
+            next_action=self.last_action,
+            is_state_terminal=done)
+
+        self.stop_episode()
+
+    def stop_episode(self):
+        self.last_state = None
```

Comment:
Are we not supporting recurrence? E.g. like here: https://github.com/chainer/chainerrl/blob/master/chainerrl/agents/ddpg.py#L468?

Reply:
This PR does not. Since the new interface of recurrent model is under review #431 , I think it is better to work on it later.